fondant/pipeline.py

```diff
@@ -0,0 +1,447 @@
+                f"Multiple component dependencies provided for component `{task.name}`. "
+                f"The current version of Fondant can only handle components with a single "
+                f"dependency. Please note that the behavior of the pipeline may be unpredictable"
+                f" or incorrect."
+            )
+        if dependencies:
+            dependencies = [dependency.name for dependency in dependencies]
+
+        task_name = task.name
+        self._graph[task_name] = {
+            "fondant_component_op": task,
+            "dependencies": dependencies,
+        }
+
+    def sort_graph(self):
+        """Sort the graph topologically based on task dependencies."""
+        sorted_graph = []
+        visited = set()
+
+        def depth_first_traversal(node: str):
+            """
+            Perform a depth-first traversal of the graph and its dependencies.
+
+            Args:
+                node: The name of the starting node for traversal.
+            """
+            if node not in visited:
+                visited.add(node)
+                for dependency in self._graph[node]["dependencies"]:
+                    depth_first_traversal(dependency)
+                sorted_graph.append(node)
+
+        for graph_node in self._graph:
+            depth_first_traversal(graph_node)
+
+        self._graph = OrderedDict((node, self._graph[node]) for node in sorted_graph)
+
+    def _validate_pipeline_definition(self, run_id: str):
+        """
+        Validates the pipeline definition by ensuring that the input and output subsets and their
+         associated fields match and are invoked in the correct order.
+
+        Raises:
+            InvalidPipelineDefinition: If a component is trying to invoke a subset that is not
+             defined or created in previous components, or if an invoked subset's schema does not
+              match the previously created subset definition.
+            base_path: the base path where to store the pipelines artifacts
+            run_id: the run id of the component
+        """
+        load_component = True
+        load_component_name = list(self._graph.keys())[0]
+
+        # Create initial manifest
+        manifest = Manifest.create(
+            base_path=self.base_path, run_id=run_id, component_id=load_component_name
+        )
+        for _, operation_specs in self._graph.items():
+            fondant_component_op = operation_specs["fondant_component_op"]
+            component_spec = FondantComponentSpec.from_file(
+                fondant_component_op.component_spec_path
+            )
+            manifest = manifest.evolve(component_spec)
+            if not load_component:
+                # Check subset exists
+                for (
+                    component_subset_name,
+                    component_subset,
+                ) in fondant_component_op.input_subsets.items():
+                    manifest_subset = manifest.subsets
```

Comment:
The assignment to a singular name is a bit confusing. But we can probably lose the assignment completely.